Add cps2pc, a script to convert cps files to pkgconfig files

fbudin69500 opened the pull request:
The new scripts takes 2 required parameters:
* The input CPS file
* And output directory
Additionally, 2 optional paramters can be given:
* The compiler that the pkgconfig options are created for (default is `gcc`).
* A prefix to replace the value '@prefix@' given in the CPS file (default is
`${pcfiledir}/../..`).

A pkgconfig file will be created for each individual component listed in the
CPS file.
The pkgconfig file is generated for a specific compilter. If not specified,
`gcc` will be used by default. This is necessary because a certain number of
parameters contained in the CPS file need to be converted from their generic
declaration to actual flags given to the compiler. This includes definitions,
link, runtime, and compile features.

The current implementation is limited: It only support a limited number of
compilers (gcc, clang), runtimes, kernels, and compiler features. The
implementation design mostly using dictionaries should allow to extend the
implementation easily if necessary, by simply adding new keys
to the dictionaries.

Certain CPS tags are ignored as they are not supported by pkgconfig:
* Link-Location
* Link-Languages
* Configurations

fbudin69500 commented:
@mwoehlke Could you review this PR and #8 ?

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,206 @@
+
+_supported_compilers = ['gcc', 'clang']
+_default_prefix = '${pcfiledir}/../..'
+#------------------------------------------------------------------------------
```
Missing blank line before.

fbudin69500 replied:
done.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,206 @@
+#------------------------------------------------------------------------------
+def get_link_features(link_features, platform):
+    _supported_kernels = ['cygwin', 'linux', 'darwin']
+    _supported_link_features = {x: {'threads': '-lpthread'} for x in _supported_kernels}
```
While I am not trying to follow *all* of PEP-8, I do prefer to keep lines under 79 characters. This one and a number of others here are "too long".

(In fact, I think I am generally in favor of following `pycodestyle` *except* for E261, E265, E302, and E305... and I'm not convinced on E731. I'm willing to relax line length when the alternative is worse, especially for only one character over, but I don't believe that's the case for many of the over-length lines here.)

See also a34ffa1f5574.

fbudin69500 replied:
done.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,206 @@
+_default_prefix = '${pcfiledir}/../..'
+#------------------------------------------------------------------------------
+def get_link_features(link_features, platform):
+    _supported_kernels = ['cygwin', 'linux', 'darwin']
```
I want to say these would make more sense as globals, all at the top, than scattered throughout the file? Is there a strong reason not to do that?

fbudin69500 replied:
I just tried to keep these as local as possible, but I will group them as globals.

mwoehlke replied:
Understood. There's an argument to be made either way, but one reason I tend to define all this stuff up front is that it makes it easier to read and modify them. Given that was a stated goal in your commit message... :slightly_smiling_face:

fbudin69500 replied:
Ok. I moved them to global variables.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,206 @@
+    except KeyError as k:
+        raise Exception("Unknown link feature %s on %s" % (k.message, platform))
+
+def get_compile_features(compile_features, compiler):
```
Missing separator line (also many following functions).

fbudin69500 replied:
done.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,206 @@
+        # * Link-Languages
+        # * Extensions
+        # * Configurations
+        # Missing pkgconfig tag (not supported by CPS):
```
Use supplemental schema? (Depends on either #8 or #10 landing first, I suppose.)

fbudin69500 replied:
updated based on #10

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,206 @@
+            raise Exception("Unknown definition argument %s" % definition)
+    return definitions
+
+
```
Second blank line should be separator line instead. (Also occurs several times below.)

fbudin69500 replied:
done.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,206 @@
+    if compiler not in _supported_compilers:
+        raise Exception('Compiler %s not supported. Check supported compiler with --supported-compiler' % compiler)
+
+
```
Class should have `#====` separator line before, and methods should have (indented `#----`) separator lines. Also, can the class be defined before the free functions? (If "yes", there should also be a `#%%%%` separator between classes and free functions. See [cps.py](https://github.com/mwoehlke/pycps/blob/master/cps.py), especially circa [line 172](https://github.com/mwoehlke/pycps/blob/a34ffa1f5574819224f6a7ee4fb47ed255e086d4/cps.py#L172) for an example.)

All that said, this seems like an overly complicated way to handle something that would be better (and *much* more simply) handled by [`choices`](https://docs.python.org/2/library/argparse.html#choices). Accordingly, I am going to refrain from any further comment on this class.

fbudin69500 replied:
True. I simplified the code.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,206 @@
+
+    args = parser.parse_args(args)
+    package = cps.read(args.input, canonicalize=True)
+    # Fail-safe function to get an easy to understand error message.
```
A blank line before here, and before `write_targets_config`, would be nice :slightly_smiling_face:.

fbudin69500 replied:
A blank line has been added.

mwoehlke reviewed (changes requested):
Some style observations... I will try to look in more detail tomorrow; meanwhile, maybe take a look at #10? I also made an editorial pass on your commit message; please check out 75880ea99e98.

mwoehlke commented:
I still have not looked at this in detail, but while thinking about your comment on #10, I realized that this:

> Certain CPS tags are ignored as they are not supported by pkgconfig: [...] Link-Location

...seems sketchy. While I realize this is one of the fundamental problems with pkg-config, shouldn't we at least decompose a library into `-L <dir> -l <libname>` based on the Link-Location?

fbudin69500 commented:
Actually splitting the path (location or link-location) into path (given to `-L`) and filename (given to `-l`) could potentially create the following issues:
* The library could be found in a different location if a file with the same name is available in a folder specified by the CPS file that is converted, or specified by a different argument (`-L`) in the command line.
* This would make it more difficult to make this script work on new platforms such as Windows.

The chosen solution is to not use any flag and directly give the full path of the library. This should work with any version of `ld` that is reasonably recent (>=2.25 or so).

mwoehlke commented:
> The chosen solution is to not use any flag and directly give the full path of the library.

Well, sure, that's better, but I am confused then; how is Link-Location not being used?

Okay, after poking around in the code some (disclaimer: still haven't read through the PR fully), I see it is... so the comment about Link-Location not being used is wrong? I was mainly reacting to that, but if it's wrong, then that's okay...

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,211 @@
+            return "unknown"
+
+#------------------------------------------------------------------------------
+def handle_kind(kind, location, link_location):
```
FYI, `link_location` will be `None` iff `location` is `None`:

https://github.com/mwoehlke/pycps/blob/568ea872ff3196e9cdaa9b3f338f6565e875b985/cps.py#L97-L98

I'm not sure you even need to pass `location` to this. (CPS doesn't work this way, partly as an artifact of being "inspired" by CMake, but with PyCPS, you will only very rarely use `location` instead of `link_location`, and that almost exclusively on Windows.)

On a different note, "handle" is just about my least favorite word to use in function names...

fbudin69500 replied:
It seems reasonable to only pass `link_location` then. I simplified the code. I renamed the function with hopefully a better name.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,211 @@
+_supported_kernels = ['cygwin', 'linux', 'darwin']
+_supported_cpp_runtime_vendors = {'gnu': 'libstdc++', 'llvm': 'libc++'}
+
+# Compile definitions
```
I appreciate the good comments for grouping stuff; thanks!

fbudin69500 commented:
`link_location` was indeed not used before I updated my PR today. It now is supported.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,210 @@
+        try:
+            output = subprocess.check_output(['uname', '-s'])
+            return output.strip().lower()
+        except:
```
s.b... erm, a subprocess exception, I think? ...but I'm too lazy to look up which :slightly_smiling_face:.

fbudin69500 replied:
OSError if the command doesn't exist.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,210 @@
+            # No need to test result `ma` as regex catches everything.
+            prefix = _compilers_definition_prefixes[compiler][ma.group(1)]
+            definitions.append(prefix + ma.group(2))
+        except:
```
s.b. `KeyError`?

fbudin69500 replied:
Yes, but this should never happened. I improved the documentation and the error message.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,210 @@
+def get_platform(package):
+    try:
+        return package.platform.kernel
+    except:
```
"E722 do not use bare except"... I believe you that L97 might throw, but can't say offhand what exception...

fbudin69500 replied:
Done.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,210 @@
+def optional(extensions, attr):
+    try:
+        return extensions.data[attr]
+    except:
```
s.b. `KeyError`?

fbudin69500 replied:
Done.

mwoehlke reviewed (changes requested):
Can you please fix the `pycodestyle --ignore=E305,E265,E302,E731,E261` errors? At least the E722 ones as noted? (I'd do it myself, but it would be good for you to verify that I'm catching the right exceptions. If you want to punt on the non-E722 reports, I'll take care of the rest.)

Commit:
Add cps2pc script to convert CPS to pkgconfig

The new scripts takes 2 required parameters:
* The input CPS file
* And output directory

Additionally, 2 optional parameters can be given:
* The compiler for which the pkgconfig options are created
  (default: 'gcc').
* A prefix to replace the value '@prefix@' given in the CPS file
  (default is '${pcfiledir}/../..').

A pkgconfig file will be created for each individual component listed in
the CPS file.  The pkgconfig file is generated for a specific compiler.
If not specified, 'gcc' will be used by default. This is necessary
because a certain number of parameters contained in the CPS file need to
be converted from their generic declaration to actual flags given to the
compiler. This includes definitions, link, runtime, and compile
features.

The current implementation only supports a limited number of compilers
(gcc, clang), kernels, and compiler features. This is mostly
implemented using dictionaries, which should allow for easy extension,
as necessary, by simply adding new keys to the dictionaries.

Certain CPS tags are ignored as they are not supported by pkgconfig:
* Link-Languages
* Link-Requires
* Configurations

fbudin69500 commented:
@mwoehlke ping

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,217 @@
+            component.definitions[None], compiler)
+        pc_dict['Cflags'] += get_compile_features(
+            component.compile_features, compiler)
+        # Remove possible package name
```
The package shouldn't have itself as a `Requires`, yes? I'm not sure what this comment is trying to say; do I miss something, or is the comment wrong?

fbudin69500 replied:
Maybe the comment is misleading. What this was meant to say is that it removes the `qualifier` of the component name.

mwoehlke replied:
Ah... yes, you can probably tell from my previous comment how I was reading it different. Thanks for clarifying.

If you wind up making further changes before I take a closer look, I guess "Remove package name from possibly-qualified component names" would be clearer.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,217 @@
+
+_supported_compilers = ['gcc', 'clang']
+_supported_kernels = ['cygwin', 'linux', 'darwin']
+_supported_cpp_runtime_vendors = {'gnu': 'stdc++', 'llvm': 'c++'}
```
pkg-config really needs this? Usually the C++ standard runtime is added automatically by using the C++ compiler...

fbudin69500 replied:
No it doesn't. I was just trying to stick to CPS options, but I'll simplify the code and remove this.

mwoehlke replied:
IIRC the idea here was to identify which library a package used, in order to avoid conflicts. IOW, if my project is using libc++, and I am looking for package "Foo", don't use a "Foo" that is linked with libstdc++. Most of the platform stuff is along the lines of being able to tell if packages are ABI-compatible. AFAIK, pkg-config doesn't really have any equivalent for that, so there isn't much we can do but ignore it.

Actually, I guess that means we *could* drop the kernel/compiler stuff too, since right now it doesn't make a practical difference (unless I missed something), but I'll leave that up to you; I'm okay keeping it, also. The comment was mainly aimed at adding link flags for the C++ standard library.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,226 @@
+
+        # Handle supplemental schema
+        for element in ['Description', 'Website']:
+            pc_dict[element] = optional(package, element)
```
The RHS can be replaced by `package.data.get(element, [])`? (And drop the unnecessary `optional` function?) I'm a little confused though why `[]` is used as a fall-back; if the attribute is present, won't it be a string? Wouldn't `None` make more sense then as the fall-back? And in that case, do we even need to use `get`, or should this just be:
```python
        for element in ['Description', 'Website']:
            if element in package.data:
                pc_dict[element] = package.data[element]
```
...? (But if you prefer to rely on `remove_empty_keys` to nuke empty values later, that's okay too.)

fbudin69500 replied:
Updated.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,226 @@
+                  (component_name, e))
+            continue
+
+        pc_dict['Libs'] += ['-l%s' % x for x in component.link_libraries]
```
FYI, I'm... not actually sure that `-l` should be added here. The spec isn't clear (partly I think from no one using this part in anger, especially as it's discouraged anyway) if these have implied semantic meaning, or if they're effectively equivalent to Link-Flags, but I suspect the latter.

fbudin69500 replied:
I am not sure how else it could be implemented. Do you mean that people cannot specify full paths with this implementation? I could remove this for now and see when one actually needs it.

mwoehlke replied:
I mean, I think if Link-Libraries is non-empty, the values will *already* be e.g. `-lfoo`. (Or... since you mention it, I guess they might be full paths, also. At any rate, what I'm getting at is that I *think* you just want `pc_dict['Libs'] += component.link_libraries`.)

fbudin69500 replied:
Ok, I understand now.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,226 @@
+                get_kind_pc_dictionary(component.kind, component.link_location)
+            )
+        except Exception as e:
+            print('Skipping %s: %s kind is not supported by pkgconfig' %
```
FYI, while I'm not exactly losing sleep over whether this will work with Python 3 or not, I'd prefer to avoid *gratuitous* Python 3 incompatibilities :wink:. Hence, I have changed your `print ...` to `print(...)`.

fbudin69500 replied:
Thanks.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,226 @@
+            continue
+
+        pc_dict['Libs'] += ['-l%s' % x for x in component.link_libraries]
+        pc_dict['Libs'] += ['-l%s' % x for x in component.link_requires]
```
This doesn't seem like it would work... Link-Requires are component names; sticking `-l` in front of those seems unlikely to work well. I don't think pkg-config really has an analogue to Link-Requires; we might need to just lump these in with other Requires.

fbudin69500 replied:
I'll go with a simpler implementation that prints a warning message saying that this is not currently supported. This could be added in the future if we need to.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,226 @@
+            features.append(_supported_link_features[platform][feature])
+        return features
+    except KeyError as k:
+        raise Exception('Unknown link feature %s on %s' %
```
Failing here might be overkill... do you think we really need to make this a fatal error, or should we just warn about it and slog on?

fbudin69500 replied:
Updated the code to print out on stderr instead.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,226 @@
+        pc_dict['Cflags'] += get_compile_features(
+            component.compile_features, compiler)
+        # Remove possible package name
+        pc_dict['Requires'] = [x.split(':')[-1] for x in component.requires]
```
I guess if this is working for your use cases, it is okay for now... I wonder if we don't need to somehow be more clever though about translating *qualified* component names into pkg-config names. (For example, I doubt e.g. `OpenCV:Core` → `Core` is really what we want. But this will depend on the naming conventions in play, hence, if it's working for you, let's not worry about it too much right now.)

fbudin69500 replied:
I'll update the comment to specify that this is incomplete and subject to improvements.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,226 @@
+            prefix = _compilers_definition_prefixes[compiler][ma.groups()[0]]
+            definitions.append(prefix + ma.groups()[1])
+        except KeyError:
+            # This should never happen as the list of supported compilers
```
If it's a bug, I wonder if it's worth bothering to futz with the exception rather than just let the original exception be raised? Up to you, I guess...

fbudin69500 replied:
I think this catch would allow to debug a potential problem more easity. I would prefer to leave this code here.

mwoehlke commented on cps2pc:
```diff
@@ -0,0 +1,226 @@
+                    ma.group(5))
+                continue
+            raise Exception(
+                'Unknown link feature %s on %s' % (k.message, compiler))
```
See comment in `get_link_features`, but anyway, you wrote "link" here :slightly_smiling_face:.

fbudin69500 replied:
Updated.

mwoehlke commented:
I pushed some changes (mostly improving comments and how warnings are reported). Please check that everything still looks okay for you. If yes, I believe this is ready.

Thanks!

fbudin69500 commented:
Great, thanks! Please merge whenever you get a chance. (@jamiesnape : This should be merged anytime now that @mwoehlke is done reviewing this PR).
